torch_glow/src/PyTorchFileLoader.cpp

```diff
@@ -15,30 +15,188 @@
+      at::Symbol::fromQualString("glow::LoaderFusionPass");
+  return fusionSymbol;
+}
+
+/// Local thread variable, used in PyTorch Custom Fusion pass.
+static thread_local LocalFusionFunction localFusionInfo;
+
+/// Convert JIT Graph into Glow Function.
+llvm::Error convertJitGraphToGlowFunction(
+    torch::jit::Stack &stack, torch::jit::Graph &graph, glow::Function &f,
+    std::vector<glow::Placeholder *> &inputPlaceholders,
+    std::vector<glow::Placeholder *> &outputPlaceholders) {
+  const auto &graphInputs = graph.inputs();
+  const auto numInputs = graphInputs.size();
+  auto inputs = torch::jit::last(stack, numInputs);
+
+  // Load JIT Graph into Glow Function.
+  RETURN_IF_ERR(PyTorchModelLoader::loadJITGraph(
+      f, graph, inputs, inputPlaceholders, outputPlaceholders));
+
+  // Remove from stack input parameters.
+  torch::jit::drop(stack, numInputs);
+  // Lookup placeholders for the output shapes.
+  for (auto *ph : outputPlaceholders) {
+    std::vector<int64_t> sizes;
+    for (auto size : ph->dims()) {
+      sizes.push_back(static_cast<int64_t>(size));
+    }
+    // Create an empty tensor with the correct shape.
+    auto ptT = at::empty(sizes);
+    auto var = torch::autograd::make_variable(ptT);
+    // Add to stack output parameter.
+    stack.push_back(at::IValue(var));
+  }
+
+  return llvm::Error::success();
+}
+
+/// Runs Module forward pass, triggers custom fusion pass if local thread
+/// Glow function is set.
+llvm::Error
+evaluateModuleGraph(std::shared_ptr<torch::jit::script::Module> &module,
+                    std::vector<torch::jit::IValue> &inputs) {
+  try {
+    module->forward(inputs);
+  } catch (const std::exception &x) {
+    RETURN_ERR(x.what());
+  }
+  return llvm::Error::success();
+}
+
+/// Helper struct, which on constructor registers custom fusion pass
+/// and operator.
+struct RegisterCustomFusionPass {
+  RegisterCustomFusionPass() {
+    auto options = c10::OperatorOptions();
+    options.setAliasAnalysis(at::AliasAnalysisKind::PURE);
+    torch::jit::RegisterOperators op({torch::jit::Operator(
+        getFusionSymbol(),
+        [](const torch::jit::Node *node) {
+          return [node](torch::jit::Stack &stack) {
+            // Get JIT Graph.
+            auto graph = node->g(at::attr::Subgraph);
+            auto err = convertJitGraphToGlowFunction(
+                stack, *graph, *localFusionInfo.function,
+                *localFusionInfo.inputPlaceholders,
+                *localFusionInfo.outputPlaceholders);
+            if (static_cast<bool>(err)) {
+              throw std::invalid_argument(llvm::toString(std::move(err)));
+            }
+            return 0;
+          };
+        },
+        options)});
+
+    torch::jit::RegisterPass pass([&](std::shared_ptr<torch::jit::Graph> &g) {
+      // Trigger custom fusion pass only if local thread Glow Function is set.
+      if (localFusionInfo.function) {
+        // Fuse all linear operators.
+        FuseLinear(g);
```

Comment:
I guess backends would generally benefit from these canonicalization passes living in Pytorch's core, huh? @kimishpatel this is what we discussed in our meeting

Reply:
Yeah, I have thought about moving pass into "common" file, need an agreement from code owner first.

Reply:
Moved.